TypeScript_ESLint.snb.md

```diff
@@ -2,92 +2,176 @@
 
-The purpose of the linter is to detect a set of patterns in source code that match common anti-patterns. These patterns are described in **rules**. The docs contain a [complete list of these rules](https://typescript-eslint.io/rules/).
+The purpose of a linter is to detect a set of patterns in source code that match common anti-patterns.
+ESLint, like many linters, defines each anti-pattern to search for in **rules**.
+ESLint's docs contain a [list of core ESLint rules](https://eslint.org/docs/rules) available for all JavaScript projects.
+TypeScript ESlint provides an additional [list of rules specific to TypeScript projects](https://typescript-eslint.io/rules).
 
-To dive into the code, let's start with the definition of a simple rule. The `no-for-in-array` rule disallows iterating over an array with a for-in loop. Here is an example of this pattern:
+For example, the `no-extra-non-null-assertion` rule catches unnecessary repeated `!`s in TypeScript code.
+Here is an example of that anti-pattern:
 
-https://sourcegraph.com/github.com/cypress-io/cypress/-/blob/packages/driver/src/cypress/proxy-logging.ts?L13
+```ts
+const scream =
+  Math.random() > 0.5 ? "What's your favorite scary movie?" : undefined;
 
-This anti-pattern is quite widespread, because it feels intuitive to use the for-in syntax to traverse an array. Doing so, however, may visit the array elements out-of-order. Instead `array.forEach` is recommended.
+// Multiple !!s are used, but only one is needed
+console.log(scream!!!.toUpperCase());
+```
 
-Let's dive into the definition of this rule in source. It is short enough to include here in its entirety:
+Let's dig into how we can create a rule to ban those unnecessary repeated `!`s.
 
-https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-for-in-array.ts
+## Declaring a Rule
 
-The rule definition contains metadata at the top and the actual logic of the rule below. Let's zoom in on that logic:
+ESLint rules are declared as objects that have two properties:
 
-https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-for-in-array.ts?L22-41
+- `meta`: Rule metadata such as configuration options and messages it may report
+- `create`: A function that creates the functions to be called by ESLint at runtime
 
-There may be some unfamiliar syntax for some here. This code might look like it's defining a function called `ForInStatement` within an object declaration, but it's really just a terse variant of declaring a function value for an object key, like this:
+We're going to dive into the definition of the `no-extra-non-null-assertion` rule in source.
+It is short enough to include here in its entirety:
 
-```typescript
-{
-    ForInStatement: (node): void => { ... }
-}
-```
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-extra-non-null-assertion.ts
```

Comment:
@beyang the existing links all point to the latest `main`. Should we change them to specific commits? Strong vote in favor from me to stay accurate.

Reply:
Yep, this sounds good, happy to do that.